pyrit/analytics/result_analysis.py

```
@@ -0,0 +1,75 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import DefaultDict, Optional
+
+from pyrit.models import AttackOutcome, AttackResult
+
+
+@dataclass
+class AttackStats:
+    success_rate: Optional[float]
+    total_decided: int
+    successes: int
+    failures: int
+    undetermined: int
+
+
+def _compute_stats(successes: int, failures: int, undetermined: int) -> AttackStats:
+    total_decided = successes + failures
+    success_rate = successes / total_decided if total_decided > 0 else None
+    return AttackStats(success_rate, total_decided, successes, failures, undetermined)
+
+
+def analyze_results(attack_results: list[AttackResult]) -> dict:
+    """
+    Analyze a list of AttackResult objects and return overall and grouped statistics.
+    """
+    if not attack_results:
+        empty_stats = AttackStats(None, 0, 0, 0, 0)
+        return {
+            "Overall": empty_stats,
+            "By_attack_identifier": {},
+        }
+
+    overall_counts: DefaultDict[str, int] = defaultdict(int)
+    by_type_counts: DefaultDict[str, DefaultDict[str, int]] = defaultdict(lambda: defaultdict(int))
+
+    for attack in attack_results:
+        if not isinstance(attack, AttackResult):
+            continue
```

Let's raise an exception because we don't want to ignore things when the caller expects that it will process all.